File: Leetcode/cpp/First Missing Positive.cc
class Solution {
public:
    int firstMissingPositive(int A[], int n) {
        for (int i = 0; i < n; i++) {
            A[i]--;
        }
        
        for (int i = 0; i < n; i++) {
            while (A[i] != i && A[i] >= 0 && A[i] < n && A[i] != A[A[i]]) {
                assert(A[i] >= 0);
                swap(A[i], A[A[i]]);
            }
        }
        for (int i = 0; i < n; i++) {
            if (A[i] != i) {
                return i + 1;
            }
        }
        return n + 1;
    }
};

File: Leetcode/python/First Missing Positive.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def firstMissingPositive(self, A):
        n = len(A)
        for i in xrange(n):
            while A[i] - 1 != i:
                a = A[i]
                if a - 1 < 0 or a - 1 >= n or A[a - 1] == a:
                    break
                A[a - 1], A[i] = A[i], A[a - 1]
        for i, a in enumerate(A):
            if a != i + 1:
               return i + 1
        else:
            return n + 1
